JavaScript-Runtime-(JSRT)-Overview.md

```diff
@@ -85,9 +85,117 @@ All other supported ES2015 features do not require any setup to run in JSRT.
+
+```cpp
+// a native function for logging
+JsValueRef CALLBACK LogCB(JsValueRef callee, bool isConstructCall, JsValueRef *arguments, unsigned short argumentCount, void *callbackState)
+{
+	for (unsigned int index = 1; index < argumentCount; index++)
+	{
+		if (index > 1)
+		{
+			wprintf(L" ");
+		}
+		JsValueRef stringValue;
+		JsConvertValueToString(arguments[index], &stringValue);
+		const wchar_t *string;
+		size_t length;
+		JsStringToPointer(stringValue, &string, &length);
+		wprintf(L"%s", string);
+	}
+	wprintf(L"\n");
+	return JS_INVALID_REFERENCE;
+}
+
+JsValueRef console, logFunc, global;
+JsPropertyIdRef consolePropId, logPropId;
+// create console object, log function, and set log function as property of console
+JsCreateObject(&console);
+JsCreateFunction(LogCB, nullptr, &logFunc);
+JsGetPropertyIdFromName(L"log", &logPropId);
+JsSetProperty(console, logPropId, logFunc, true);
+// set console as property of global object
+JsGetGlobalObject(&global);
+JsGetPropertyIdFromName(L"console", &consolePropId);
+JsSetProperty(global, consolePropId, console, true);
+```
+
+Now in JavaScript, you can do,
+```js
+console.log('Hello world');
+```
+
+If you need to expose an object that hosts data not directly exposed to JS, you can create an external object via `JsCreateExternalObject`.
+
+### Call JavaScript functions from native
+
+You can call JavaScript functions by calling `JsRun/JsRunScript`, or alternatively you can get the native reference of your JS function and then use `JsCallFunction` to call it. For example, to call `func` defined in global scope in JS,
+
+```cpp
+JsValueRef func, funcPropId, global, undefined, result;
+JsGetGlobalObject(&global);
+JsGetUndefinedValue(&undefined);
+JsValueRef args[] = { undefined };
+JsGetPropertyIdFromName(L"func", &funcPropId);
+JsGetProperty(global, funcPropId, &func);
+// note that args[0] is thisArg of the call; actual args start at index 1
+JsCallFunction(func, args, 1, &result);
+``` 
+
+### Script serialization with lazy source loading  
+
+To support startup and execution efficiency, JSRT APIs provide the capability of pre-parsing, generating syntax trees and caching the bytecode for scripts. It is especially useful for server scenarios where the same script could get executed for thousands of times.
+
+The basic idea is to use variants of `JsSerializeScript` to parse and store the bytecode and then use `JsParse/RunSerializedScript` to parse or run the bytecode. Source is typically not needed for running serialized code, therefore you can also lazy-load the source script using `JsRunSerialized/JsRunSerializedScriptWithCallback`.
+
+
+```cpp
+// serialize script and cache the bytecode on disk
+void serializeScript(const wchar_t *script, const wchar_t *bytecodeBufferPath) {
+    JsValueRef result;
+    unsigned long bufferSize = 0;
+    BYTE *buffer = nullptr;
+    JsSerializeScript(script, buffer, &bufferSize); // get correct buffer size
```

Comment:
nit: to make this clearer (passing `nullptr` for buffer argument to get size without filling a buffer or getting an error about buffer too small), replace `buffer` with `nullptr` in this line.